airflow-core/src/airflow/models/variable.py

```diff
@@ -230,8 +230,61 @@ def set(
         else:
             stored_value = str(value)
 
-        Variable.delete(key, session=session)
-        session.add(Variable(key=key, val=stored_value, description=description))
+        new_variable = Variable(key=key, val=stored_value, description=description)
+
+        # Perform dialect-specific upsert operation
+        dialect_name = session.get_bind().dialect.name
+
+        # Use SQLAlchemy Core for supported dialects
+        if dialect_name in ("postgresql", "mysql", "sqlite"):
```

Comment:
```suggestion
        if dialect_name in dialect_insert_map:
            # Dynamically import the insert function
            insert = import_string(dialect_insert_map[dialect_name])
```

Note: `import_string` is a airflow utils from `airflow.utils.module_loading`